packages/plugin-core/src/commands/CopyNoteLink.ts

```diff
@@ -175,13 +186,46 @@ export class CopyNoteLinkCommand extends BasicCommand<
     const editor = VSCodeUtils.getActiveTextEditor()!;
     const fname = NoteUtils.uri2Fname(editor.document.uri);
     const engine = ExtensionProvider.getEngine();
-
     const vault = PickerUtilsV2.getVaultForOpenEditor();
-    const note = NoteUtils.getNoteByFnameFromEngine({
-      fname,
-      vault,
-      engine,
-    }) as NoteProps;
+
+    if (editor.document.isDirty) {
+      this._onEngineNoteStateChangedDisposable =
+        this.engineEvents.onEngineNoteStateChanged(
+          async (noteChangeEntries: NoteChangeEntry[]) => {
+            const savedNote = noteChangeEntries.filter(
+              (entry) => entry.note.fname === fname && entry.status === "update"
+            );
+            // Received event from engine about successful save
+            if (savedNote.length > 0) {
+              await this.executeCopyNoteLink(savedNote[0].note, editor);
+              this.dispose();
+            }
+          }
+        );
+      await editor.document.save();
+      return;
+    } else if (this._onEngineNoteStateChangedDisposable) {
+      // If this is not disposed, it means we are still listening on engine state change from previous CopyNoteLink.execute command
```

Comment:
Hmm is this true though?  This command isn't implemented as a singleton, so I think concurrent calls of CopyNoteLink will result in multiple subscriptions.  When the command is run, what's invoked by vscode isn't this `execute()`, rather, it's what we wrote in `_extension.ts`:

```typescript
        sentryReportingCallback(async (args) => {
          if (args === undefined) {
            args = {};
          }
          await new CopyNoteLinkCommand(ws.getEngine()).run(args); // We'll have multiple new calls with multiple invocations.
        })
```

Reply:
Oh you're right. In that case, do we need to create new each time the key is invoked? If we instantiate outside the `registerCommand`, we can use the same instance. Otherwise multiple commands might be trying to save if the first command isn't "done" yet.

Reply:
I think that should work well